torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,731 @@
+                    out_spec
+                    if expected_input_specs is None
+                    else expected_input_specs[idx]
+                )
+                if input_arg_spec != desired_spec:
+                    # insert reshard operation
+                    def reshard_fn(local_tensor: torch.Tensor) -> torch.Tensor:
+                        return _redistribute_with_local_tensor(
+                            local_tensor,
+                            full_shape,
+                            out_spec.mesh,
+                            input_arg_spec.placements,
+                            desired_spec.placements,
+                        )
+
+                    reshard_gm = make_fx(reshard_fn)(input_arg_tensor)
+                    reshard_gm_nodes = list(reshard_gm.graph.nodes)
+                    input_node = reshard_gm_nodes[0]
+                    with graph.graph.inserting_before(node):
+                        output_node = graph.graph.graph_copy(
+                            reshard_gm.graph,
+                            val_map={
+                                input_node: input_arg,
+                            },
+                        )
+                    node.replace_input_with(input_arg, output_node)
+            if node.target in shape_adjustment_ops:
+                local_shape = compute_local_shape(
+                    full_shape, out_spec.mesh, out_spec.placements
+                )
+                node.update_arg(1, local_shape)
+
+            # convert output to local shard
+            # note that if the output is already local, we don't need to convert
+            if not hasattr(out_spec, "from_local"):
+                local_val = _to_local_shard(full_tensor, out_spec)
+                local_tensor_meta = _extract_tensor_metadata(local_val)
+                # update metadata
+                node.meta["val"] = local_val
+                node.meta["tensor_meta"] = local_tensor_meta
+
+        elif node.op == "output":
+            break
+        else:
+            raise RuntimeError(f"op code {node} not supported")
+
+    graph.graph.lint()
+    graph.recompile()
+    graph.print_readable()
+    return graph
+
+
+def partition_data_parallel(
+    graph: GraphModule,
+    model: nn.Module,
+    optimizer: torch.optim.Optimizer,
+    params_buffers: Dict[str, torch.Tensor],
+    named_states: Dict[str, Any],
+    args: Tuple[Any, ...],
+    kwargs: Dict[str, Any],
+    mesh: DeviceMesh,
+    parallel_style: DataParallelStyle,
+) -> GraphModule:
+    """
+    The entry point function to partition the graph to data parallel
+    graph, it also shard/replicate the model parameters and optimizer
+    states to DTensors.
+    """
+    num_params_buffers = len(params_buffers)
+    flattened_states = pytree.tree_flatten(named_states)[0]
+    num_states = len(flattened_states)
+    # 1. First build up data parallel strategies for the whole graph
+    strategy_map = build_data_parallel_strategies(
```

Comment:
For now, we assume batch dimension is alway dim-0?

How do we get the batch dimension in general if it is not dim-0?

Reply:
Yeah usually when doing ddp/fsdp we have this implicit assumption that batch dim is 0. However for those cases that aren't batch dim = 0, in batch dim analyzer we have a field to record batch dim, it currently infer from the first input directly, but if user pass in a custom batch dim, we might can adjust that. It's really hard to know the batch dim if user don't provide a hint though, I think this can be a field of `DataParallelMode`